File: Frontend/lib/core/models/vehicle.dart
class Vehicle {
  final int id;

  final String registrationNumber;
  final String vehicleModel;
  final String? vehicleColor;

  final String vehicleCategory;
  final String vehicleSubtype;
  final String? vehicleTypeSpecified;
  final int seatingCapacity;

  final bool isActive;
  final bool rcVerified;

  const Vehicle({
    required this.id,
    required this.registrationNumber,
    required this.vehicleModel,
    this.vehicleColor,
    required this.vehicleCategory,
    required this.vehicleSubtype,
    this.vehicleTypeSpecified,
    required this.seatingCapacity,
    required this.isActive,
    required this.rcVerified,
  });

  factory Vehicle.fromJson(Map<String, dynamic> json) {
    return Vehicle(
      id: json['id'] as int,
      registrationNumber:
          json['registration_number'] as String,
      vehicleModel:
          json['vehicle_model'] as String,
      vehicleColor:
          json['vehicle_color'] as String?,

      vehicleCategory:
          json['vehicle_category'] as String,
      vehicleSubtype:
          json['vehicle_subtype'] as String,
      vehicleTypeSpecified:
          json['vehicle_type_specified'] as String?,
      seatingCapacity:
          json['seating_capacity'] as int,

      isActive:
          json['is_active'] as bool? ?? false,
      rcVerified:
          json['rc_verified'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'registration_number': registrationNumber,
      'vehicle_model': vehicleModel,
      'vehicle_color': vehicleColor,

      'vehicle_category': vehicleCategory,
      'vehicle_subtype': vehicleSubtype,
      'vehicle_type_specified': vehicleTypeSpecified,
      'seating_capacity': seatingCapacity,

      'is_active': isActive,
      'rc_verified': rcVerified,
    };
  }

  Vehicle copyWith({
    int? id,
    String? registrationNumber,
    String? vehicleModel,
    String? vehicleColor,

    String? vehicleCategory,
    String? vehicleSubtype,
    String? vehicleTypeSpecified,
    int? seatingCapacity,

    bool? isActive,
    bool? rcVerified,
  }) {
    return Vehicle(
      id: id ?? this.id,
      registrationNumber:
          registrationNumber ?? this.registrationNumber,
      vehicleModel:
          vehicleModel ?? this.vehicleModel,
      vehicleColor:
          vehicleColor ?? this.vehicleColor,

      vehicleCategory:
          vehicleCategory ?? this.vehicleCategory,
      vehicleSubtype:
          vehicleSubtype ?? this.vehicleSubtype,
      vehicleTypeSpecified:
          vehicleTypeSpecified ?? this.vehicleTypeSpecified,
      seatingCapacity:
          seatingCapacity ?? this.seatingCapacity,

      isActive:
          isActive ?? this.isActive,
      rcVerified:
          rcVerified ?? this.rcVerified,
    );
  }
}

File: Backend/app/db/schemas/vehicle.py
from pydantic import BaseModel, Field, model_validator


# ═══════════════════════════════════════════════════════════════
# VEHICLE RESPONSE
# ═══════════════════════════════════════════════════════════════


class VehicleResponse(BaseModel):
    """
    Vehicle information returned to the authenticated user.

    Sensitive RC data is never returned.
    """

    id: int

    registration_number: str

    vehicle_model: str

    vehicle_color: str | None = None

    # ─── Vehicle Classification ─────────────────────────────────

    vehicle_category: str

    vehicle_subtype: str

    # Populated only when the vehicle type is custom/other.
    vehicle_type_specified: str | None = None

    # ─── Capacity ───────────────────────────────────────────────

    # Total verified passenger/seating capacity.
    seating_capacity: int

    # ─── User Vehicle Status ────────────────────────────────────

    is_active: bool

    rc_verified: bool


# ═══════════════════════════════════════════════════════════════
# VEHICLE LIST RESPONSE
# ═══════════════════════════════════════════════════════════════


class VehicleListResponse(BaseModel):
    """
    Response containing all vehicles associated
    with the authenticated user.
    """

    vehicles: list[VehicleResponse]


# ═══════════════════════════════════════════════════════════════
# VEHICLE UPDATE REQUEST
# ═══════════════════════════════════════════════════════════════


class VehicleUpdateRequest(BaseModel):
    """
    Update vehicle metadata.

    Registration number, classification and seating capacity
    cannot be changed through this endpoint because they belong
    to the verified vehicle identity.

    If any of these permanent properties need to change,
    the vehicle should go through the appropriate verification
    process again.
    """

    vehicle_model: str | None = Field(
        default=None,
        min_length=2,
        max_length=100,
        description="Vehicle make and model",
    )

    vehicle_color: str | None = Field(
        default=None,
        max_length=50,
        description="Vehicle color",
    )


# ═══════════════════════════════════════════════════════════════
# VEHICLE STATUS UPDATE
# ═══════════════════════════════════════════════════════════════


class VehicleStatusUpdateRequest(BaseModel):
    """
    Activate or deactivate a user's vehicle association.
    """

    is_active: bool